File: ctrader/json/ctrader.js
var credentials = require('./credentials.json');
var oa = require('./OAModel.json');
var pt = require('./payloadTypes.json');
uid = (i => () => 'cm_id_' + i++)(1);

// connection, authentication and program lifecycle
var { accountId: ctidTraderAccountId, accessToken } = credentials;
var ws = new WebSocket('wss://live.ctraderapi.com:5036'); // or 'demo.ctraderapi.com:5036'
function onResp(message) {
	console.log('main messages come here', message.payload);
}
function main() {
	console.log('ready to do something');
}
ws.onopen = function (e) {
	console.log('connected', e);
	var clientMsg = {
		clientMsgId: uid(),
		payloadType: pt.req.ApplicationAuth,
		payload: {
			clientId: credentials.clientId,
			clientSecret: credentials.clientSecret
		}
	};
	ws.send(JSON.stringify(clientMsg));
};
ws.onmessage = function (e) {
	var serverMsg = JSON.parse(e.data);
	var { payloadType } = serverMsg;
	if (payloadType === pt.res.ApplicationAuth) {
		console.log('app auth done');
		var clientMsg = {
			clientMsgId: uid(),
			payloadType: pt.req.AccountAuth,
			payload: {ctidTraderAccountId, accessToken},
		};
		ws.send(JSON.stringify(clientMsg));
		return;
	}
	if (payloadType === pt.res.AccountAuth) {
		console.log('account auth done');
		main();
		return;
	}
	if (payloadType === pt.res.Error) {
		console.log('server sent error message', serverMsg.payload);
		return;
	}
	if (payloadType === pt.common.Heartbeat) {
		console.log('heartbeat event');
		return;
	}
	onResp(serverMsg);
};
ws.onerror = function (e) {
	console.log('websocket error', e);
};
ws.onclose = function (e) {
	console.log('disconnected', e.code, e.reason, e.wasClean);
};


// get symbol list
var fs = require('fs');
function onResp(message) {
	if (message.payloadType !== pt.res.SymbolsList) return;
	var syms = message.payload.symbol;
	fs.writeFileSync('syms.json', JSON.stringify(syms,null,2));
}
function main() {
	var clientMsg = {
		clientMsgId: uid(),
		payloadType: pt.req.SymbolsList,
		payload: { ctidTraderAccountId, accessToken,
			includeArchivedSymbols: false,
		}
	};
	ws.send(JSON.stringify(clientMsg));
}


// get timeseries data
var fs = require('fs');
function onResp(message) {
	if (message.payloadType !== pt.res.GetTrendbars) return;
	var headers = ['timestamp','open','high','low','close','volume'];
	var bars = message.payload.trendbar.map(bar => {
		var timestamp = bar.utcTimestampInMinutes * 60; // seconds
		var open = (bar.low + bar.deltaOpen) / 100_000;
		var high = (bar.low + bar.deltaHigh) / 100_000;
		var low = bar.low / 100_000;
		var close = (bar.low + bar.deltaClose) / 100_000;
		return [timestamp, open, high, low, close, bar.volume];
	});
	var out = [headers, ...bars];
	fs.writeFileSync('bars.json', JSON.stringify(out,null,2));
}
function main() {
	var clientMsg = {
		clientMsgId: uid(),
		payloadType: pt.req.GetTrendbars,
		payload: { ctidTraderAccountId, accessToken,
			symbolId: 41, // XAUUSD
			period: OAModel.ProtoOATrendbarPeriod.D1,
			fromTimestamp: Date.UTC(2025,0,1),
			toTimestamp: Date.UTC(2025,0,15),
		}
	};
	ws.send(JSON.stringify(clientMsg));
}


// place order
function onResp(message) {
	if (message.payloadType !== pt.event.Execution) return;
	console.log('order successfully placed');
	console.log('order id:', message.payload.order.orderId);
}
function main() {
	var clientMsg = {
		clientMsgId: uid(),
		payloadType: pt.req.NewOrder,
		payload: { ctidTraderAccountId, accessToken,
			orderType: oa.OrderType.LIMIT,
			tradeSide: oa.TradeSide.BUY,
			symbolId: 41, // XAUUSD
			limitPrice: 2700.34,
			volume: 100, // equals 0.01 lot (note: for EURUSD 100_000 equals 0.01 lot)
			stopLoss: 2600.34,
			takeProfit: 2750.34,
		}
	};
	ws.send(JSON.stringify(clientMsg));
}


// place order - relative sl,tp
function calcRelativeByPrice(price, pipDistance) {
	const decimalCount = price.toString().split('.').at(-1).length;
	const pip = +('0.' + '1'.padStart(decimalCount, '0'));
	return Math.round(pipDistance * pip * 100_000);
}
function calcRelativeByPip(pip, pipDistance) {
	return Math.round(pipDistance * pip * 100_000);
}
function onResp(message) {
	if (message.payloadType !== pt.event.Execution) return;
	console.log(message.payload);
}
function main() {
	var order = {
		orderType: oa.OrderType.LIMIT,
		tradeSide: oa.TradeSide.BUY,
		symbolId: 41, // XAUUSD
		limitPrice: 3327.45,
		volume: 100,
	};
	order.relativeStopLoss = calcRelativeByPrice(order.limitPrice, 200);
	order.relativeTakeProfit = calcRelativeByPrice(order.limitPrice, 350);
	// or
	order.relativeStopLoss = calcRelativeByPip(0.01, 200);
	order.relativeTakeProfit = calcRelativeByPip(0.01, 350);
	var clientMsg = {
		clientMsgId: uid(),
		payloadType: pt.req.NewOrder,
		payload: {ctidTraderAccountId, accessToken, ...order}
	};
	ws.send(JSON.stringify(clientMsg));
}

File: ctrader/json/ctrader.py
import json
import asyncio as aio
from types import SimpleNamespace
from websockets.asyncio.client import connect # pip install websockets

# utils
parsejson = lambda s: json.loads(s, object_hook=lambda d: SimpleNamespace(**d))
def readfile(path):
	with open(path) as f:
		return parsejson(f.read())

creds = readfile('./credentials.json')
oa = readfile('./OAModel.json')
pt = readfile('./payloadTypes.json')


# set up a function for constructing client messages
uid_counter = 0
def construct_client_msg(payloadType, fields={}):
	global uid_counter
	uid_counter += 1
	client_msg = {
		'clientMsgId': f'cm_id_{uid_counter}',
		'payloadType': payloadType,
		'payload': {
			'ctidTraderAccountId': creds.accountId,
			'accessToken': creds.accessToken,
			**fields
		}
	}
	return json.dumps(client_msg)


async def main():
	async with connect('wss://live.ctraderapi.com:5036') as ws:
		print('Application authentication')
		# we don't use the `construct_client_msg()` function for the first message,
		# because its payload is different from the rest of the messages
		client_msg = {
			'clientMsgId': f'cm_id_{uid_counter}',
			'payloadType': pt.req.ApplicationAuth,
			'payload': {
				'clientId': creds.clientId,
				'clientSecret': creds.clientSecret
			}
		}
		await ws.send(json.dumps(client_msg))
		server_msg = parsejson(await ws.recv())
		if server_msg.payloadType != pt.res.ApplicationAuth:
			print('Application authentication failed')
		
		print('Account authentication')
		client_msg = construct_client_msg(pt.req.AccountAuth)
		await ws.send(client_msg)
		server_msg = parsejson(await ws.recv())
		if server_msg.payloadType != pt.res.AccountAuth:
			print('Account authentication failed')
		
		print('Getting symbols list')
		client_msg = construct_client_msg(pt.req.SymbolsList)
		await ws.send(client_msg)
		server_msg = parsejson(await ws.recv())
		if server_msg.payloadType != pt.res.SymbolsList:
			print('Getting symbols failed')
		
		print('Saving symbol list to file')
		symbols = [sym.__dict__ for sym in server_msg.payload.symbol]
		with open('symbols.json', 'w', encoding='utf-8') as f:
			json.dump(symbols, f, ensure_ascii=False, indent=2)
		
aio.run(main())
